### PR TITLE
Update DirectML from 1.14.1 to 1.15.0

### PR DESCRIPTION
**Pull request by smk2007**
Update DirectML from 1.14.1 to 1.15.0

### CODE REVIEW

**Review by sumitsays — DISMISSED** [review]

**Comment by martinb35** [comment]
do we really have to change it in 5 places? feels error prone. I don't want to over-engineer this, but is there a way to put it in only one place?

**Comment by fdwr** [comment]
> do we really have to change it in 5 places? feels error prone. I don't want to over-engineer this, but is there a way to put it in only one place?

Looks more error prone than it's actually ever been. In VS code, it's a single Ctrl+Shift+F, and then replace x with y :).

**Review by fdwr — APPROVED** [review]

**Review by sumitsays — APPROVED** [review]

**Review by fdwr — APPROVED** [review]

**Review by sumitsays — APPROVED** [review]

File: onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/DmlOperatorResize.cpp [hunk 1]
```diff
@@ -307,25 +302,35 @@ class DmlOperatorResize : public DmlOperator, public ResizeHelper
...
+          operatorDesc.InputTensor = inputDescs.data();
+          operatorDesc.OutputTensor = outputDescs.data();
+          operatorDesc.InterpolationMode = interpolationMode;
+          operatorDesc.RoundingDirection = roundingDirection;
+          operatorDesc.Scales = paddedScales.data();
+          operatorDesc.DimensionCount = gsl::narrow_cast<uint32_t>(paddedScales.size());
+          operatorDesc.InputPixelOffsets = inputPixelOffsets.data();
+          operatorDesc.OutputPixelOffsets = outputPixelOffsets.data();
+          opDesc = { DML_OPERATOR_RESAMPLE2, &operatorDesc };
+        }
```
Comment (fdwr) [inline onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/DmlOperatorResize.cpp:332]: Observe intra-file consistency 👀 - see a few lines above uses 4-column indents, and so 4-column indents here too.

```suggestion
        DML_OPERATOR_DESC opDesc = {};
        if (opsetVersion >= 18)
        {
            // Restrict this change to Resize18 and Resize19.
            const int antialiased = kernelCreationContext.GetOptionalAttribute<int>(AttrName::Antialiased, 0);
            DML_RESAMPLE3_OPERATOR_DESC operatorDesc = {};
            operatorDesc.Antialiased = static_cast<BOOL>(antialiased);
            operatorDesc.InputTensor = inputDescs.data();
            operatorDesc.OutputTensor = outputDescs.data();
            operatorDesc.InterpolationMode = interpolationMode;
            operatorDesc.RoundingDirection = roundingDirection;
            operatorDesc.Scales = paddedScales.data();
            operatorDesc.DimensionCount = gsl::narrow_cast<uint32_t>(paddedScales.size());
            operatorDesc.InputPixelOffsets = inputPixelOffsets.data();
            operatorDesc.OutputPixelOffsets = outputPixelOffsets.data();
            opDesc = { DML_OPERATOR_RESAMPLE3, &operatorDesc };
        }
        else
        {
            DML_RESAMPLE2_OPERATOR_DESC operatorDesc = {};
            operatorDesc.InputTensor = inputDescs.data();
            operatorDesc.OutputTensor = outputDescs.data();
            operatorDesc.InterpolationMode = interpolationMode;
            operatorDesc.RoundingDirection = roundingDirection;
            operatorDesc.Scales = paddedScales.data();
            operatorDesc.DimensionCount = gsl::narrow_cast<uint32_t>(paddedScales.size());
            operatorDesc.InputPixelOffsets = inputPixelOffsets.data();
            operatorDesc.OutputPixelOffsets = outputPixelOffsets.data();
            opDesc = { DML_OPERATOR_RESAMPLE2, &operatorDesc };
        }
```

**Review by fdwr — APPROVED** [review]

File: onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/DmlOperatorResize.cpp [hunk 2]
```diff
@@ -307,25 +302,33 @@ class DmlOperatorResize : public DmlOperator, public ResizeHelper
...
+          operatorDesc.InputTensor = inputDescs.data();
+          operatorDesc.OutputTensor = outputDescs.data();
+          operatorDesc.InterpolationMode = interpolationMode;
+          operatorDesc.RoundingDirection = roundingDirection;
+          operatorDesc.Scales = paddedScales.data();
+          operatorDesc.DimensionCount = gsl::narrow_cast<uint32_t>(paddedScales.size());
+          operatorDesc.InputPixelOffsets = inputPixelOffsets.data();
+          operatorDesc.OutputPixelOffsets = outputPixelOffsets.data();
+          opDesc = { DML_OPERATOR_RESAMPLE2, &operatorDesc };
+        #endif
```
Comment (fdwr) [inline onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/DmlOperatorResize.cpp:330]: Fix rogue indentation again.

```suggestion
        DML_OPERATOR_DESC opDesc = {};
        #if FALSE
            const int antialiased = kernelCreationContext.GetOptionalAttribute<int>(AttrName::Antialiased, 0);
            DML_RESAMPLE3_OPERATOR_DESC operatorDesc = {};
            operatorDesc.Antialiased = static_cast<BOOL>(antialiased);
            operatorDesc.InputTensor = inputDescs.data();
            operatorDesc.OutputTensor = outputDescs.data();
            operatorDesc.InterpolationMode = interpolationMode;
            operatorDesc.RoundingDirection = roundingDirection;
            operatorDesc.Scales = paddedScales.data();
            operatorDesc.DimensionCount = gsl::narrow_cast<uint32_t>(paddedScales.size());
            operatorDesc.InputPixelOffsets = inputPixelOffsets.data();
            operatorDesc.OutputPixelOffsets = outputPixelOffsets.data();
            opDesc = { DML_OPERATOR_RESAMPLE3, &operatorDesc };
        #else
            DML_RESAMPLE2_OPERATOR_DESC operatorDesc = {};
            operatorDesc.InputTensor = inputDescs.data();
            operatorDesc.OutputTensor = outputDescs.data();
            operatorDesc.InterpolationMode = interpolationMode;
            operatorDesc.RoundingDirection = roundingDirection;
            operatorDesc.Scales = paddedScales.data();
            operatorDesc.DimensionCount = gsl::narrow_cast<uint32_t>(paddedScales.size());
            operatorDesc.InputPixelOffsets = inputPixelOffsets.data();
            operatorDesc.OutputPixelOffsets = outputPixelOffsets.data();
            opDesc = { DML_OPERATOR_RESAMPLE2, &operatorDesc };
        #endif
```

File: onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/DmlOperatorResize.cpp [hunk 3]
```diff
@@ -307,25 +302,33 @@ class DmlOperatorResize : public DmlOperator, public ResizeHelper
...
-        operatorDesc.DimensionCount = gsl::narrow_cast<uint32_t>(paddedScales.size());
-        operatorDesc.InputPixelOffsets = inputPixelOffsets.data();
-        operatorDesc.OutputPixelOffsets = outputPixelOffsets.data();
-#if DML_TARGET_VERSION >= 0x6400
-        DML_OPERATOR_DESC opDesc = { DML_OPERATOR_RESAMPLE3, &operatorDesc };
-#else
-        DML_OPERATOR_DESC opDesc = { DML_OPERATOR_RESAMPLE2, &operatorDesc };
-#endif
+        DML_OPERATOR_DESC opDesc = {};
+        #if FALSE
```
Comment (fdwr) [inline onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/DmlOperatorResize.cpp:306]: This really warrants a comment / todo / issue link.

**Review by fdwr — APPROVED** [review]

File: winml/test/api/LearningModelSessionAPITest.cpp [hunk 4]
```diff
@@ -2117,15 +2117,65 @@ static void ModelBuilding_HannWindow() {
...
+  LearningModelSession session(model, device);
+  LearningModelBinding binding(session);
+
+  binding.Bind(L"Input", TensorFloat::CreateFromArray(input_shape, input));
+
+  // Evaluate
+  auto result = session.Evaluate(binding, L"");
+
+  // Check results
+ // constexpr float error_threshold = .001f;
```
Comment (fdwr) [inline winml/test/api/LearningModelSessionAPITest.cpp:2169]:
```suggestion
  // constexpr float error_threshold = .001f;
```

File: winml/test/api/LearningModelSessionAPITest.cpp [hunk 5]
```diff
@@ -2117,15 +2117,65 @@ static void ModelBuilding_HannWindow() {
...
+  auto y_tensor = result.Outputs().Lookup(L"Output").as<TensorFloat>();
+  auto y_ivv = y_tensor.GetAsVectorView();
+  for (uint32_t i = 0; i < y_ivv.Size(); i++) {
+    if (i % 5 == 0) {
+      printf("\n");
+    }
+    printf("%f, ", y_ivv.GetAt(i));
+    //WINML_EXPECT_TRUE(abs(y_ivv.GetAt(i) - expected[i]) < error_threshold);
```
Comment (fdwr) [inline winml/test/api/LearningModelSessionAPITest.cpp:2177]: Were you debugging something here and forgot to re-enable it? As-is, the test doesn't appear to actually validate anything.

**Review by fdwr — DISMISSED** [review]
Conditional approval, depending on your answer to LearningModelSessionAPITest.cpp.

File: winml/test/api/LearningModelSessionAPITest.cpp (hunk 5 above)
Comment (smk2007) [reply winml/test/api/LearningModelSessionAPITest.cpp:2177]: oh whoops... that file shouldnt be there!

File: onnxruntime/core/providers/dml/DmlExecutionProvider/src/External/DirectMLHelpers/ApiHelpers.h [hunk 6]
```diff
@@ -46,7 +49,7 @@ struct ActivationOperatorDesc
         case DML_OPERATOR_ACTIVATION_CELU: return { activationType, &params.celu };
         case DML_OPERATOR_ACTIVATION_HARDMAX: return { activationType, &params.hardmax };
         case DML_OPERATOR_ACTIVATION_HARDMAX1: return { activationType, &params.hardmax1 };
-        case DML_OPERATOR_ACTIVATION_HARD_SIGMOID: return { activationType, &params.sigmoid };
```
Comment (sumitsays) [inline onnxruntime/core/providers/dml/DmlExecutionProvider/src/External/DirectMLHelpers/ApiHelpers.h:49]: Wondering how this typo lived this long un-noticed. Thank you for fixing this.

File: onnxruntime/core/providers/dml/DmlExecutionProvider/src/ApiTraits.cpp [hunk 7]
```diff
@@ -241,6 +241,7 @@ DML_OPERATOR_TYPE ApiTraits::StringifyHelpers::FromString(std::string_view value
         {"DML_OPERATOR_ACTIVATION_SWISH", DML_OPERATOR_ACTIVATION_SWISH},
         {"DML_OPERATOR_ACTIVATION_HARD_SWISH", DML_OPERATOR_ACTIVATION_HARD_SWISH},
         {"DML_OPERATOR_RESAMPLE2", DML_OPERATOR_RESAMPLE2},
+        {"DML_OPERATOR_RESAMPLE3", DML_OPERATOR_RESAMPLE3},
```
Comment (sumitsays) [inline onnxruntime/core/providers/dml/DmlExecutionProvider/src/ApiTraits.cpp:244]: We should also have entries for Fold and Unfold.

Comment (fdwr) [reply onnxruntime/core/providers/dml/DmlExecutionProvider/src/ApiTraits.cpp:244]: Interesting. I see them below, but not here:

```
template <>
struct OperatorDescTraits<DML_FOLD_OPERATOR_DESC>
{
    static constexpr DML_OPERATOR_TYPE Type = DML_OPERATOR_FOLD;
};

template <>
struct OperatorDescTraits<DML_UNFOLD_OPERATOR_DESC>
{
    static constexpr DML_OPERATOR_TYPE Type = DML_OPERATOR_UNFOLD;
};
```

File: onnxruntime/core/providers/dml/OperatorAuthorHelper/OperatorHelper.cpp [hunk 8]
```diff
@@ -1857,30 +1857,30 @@ namespace OperatorHelper
...
         const uint32_t dimCount = gsl::narrow_cast<uint32_t>(m_blockShape.size());
-        m_dilations = {dimCount, 1};
-        m_pads = {dimCount * 2, 0};
-        m_strides = {dimCount, 1};
+        m_dilations.resize(dimCount);
+        std::fill(m_dilations.begin(), m_dilations.end(), 1);
+        m_pads.resize(dimCount, 0);
+        std::fill(m_pads.begin(), m_pads.end(), 0);
+        m_strides.resize(dimCount, 1);
+        std::fill(m_strides.begin(), m_strides.end(), 1);
```
Comment (fdwr) [inline onnxruntime/core/providers/dml/OperatorAuthorHelper/OperatorHelper.cpp:1865]: Why repeat a fill when the resize just filled it?
```suggestion
        m_dilations.resize(dimCount, 1);
        m_pads.resize(dimCount, 0);
        m_strides.resize(dimCount, 1);
```

File: onnxruntime/core/providers/dml/OperatorAuthorHelper/OperatorHelper.cpp [hunk 9]
```diff
@@ -1857,30 +1857,30 @@ namespace OperatorHelper
         DowncastDimensions(gsl::span(shapeData), /*out*/ m_blockShape);
 
         const uint32_t dimCount = gsl::narrow_cast<uint32_t>(m_blockShape.size());
-        m_dilations = {dimCount, 1};
-        m_pads = {dimCount * 2, 0};
-        m_strides = {dimCount, 1};
+        m_dilations.resize(dimCount);
+        std::fill(m_dilations.begin(), m_dilations.end(), 1);
+        m_pads.resize(dimCount, 0);
```
Comment (fdwr) [inline onnxruntime/core/providers/dml/OperatorAuthorHelper/OperatorHelper.cpp:1862]: I was `dimCount * 2` before. Should it be `m_pads.resize(dimCount * 2, 0);`?

Comment (smk2007) [reply onnxruntime/core/providers/dml/OperatorAuthorHelper/OperatorHelper.cpp:1862]: Its correct now!

File: onnxruntime/core/providers/dml/DmlExecutionProvider/src/ApiTraits.cpp (hunk 7 above)
Comment (smk2007) [reply onnxruntime/core/providers/dml/DmlExecutionProvider/src/ApiTraits.cpp:244]: Yeah, it looks like there was another entry in the schema that was missing. It shouldnt cause a breaking change though.
Lets keep ApiTraits in sync and update with the following release of DML which has the fix.

**Review by fdwr — APPROVED** [review]

**Review by fdwr — APPROVED** [review]

**Review by sumitsays — APPROVED** [review]

**Pull request merged**